include/exadg/acoustic_conservation_equations/spatial_discretization/spatial_operator.cpp

```diff
@@ -0,0 +1,383 @@
+{
+  return *matrix_free;
+}
+
+template<int dim, typename Number>
+std::string
+SpatialOperator<dim, Number>::get_dof_name_pressure() const
+{
+  return field + dof_index_p;
+}
+
+template<int dim, typename Number>
+unsigned int
+SpatialOperator<dim, Number>::get_dof_index_pressure() const
+{
+  return matrix_free_data->get_dof_index(get_dof_name_pressure());
+}
+
+template<int dim, typename Number>
+std::string
+SpatialOperator<dim, Number>::get_dof_name_velocity() const
+{
+  return field + dof_index_u;
+}
+
+template<int dim, typename Number>
+unsigned int
+SpatialOperator<dim, Number>::get_dof_index_velocity() const
+{
+  return matrix_free_data->get_dof_index(get_dof_name_velocity());
+}
+
+template<int dim, typename Number>
+unsigned int
+SpatialOperator<dim, Number>::get_quad_index_pressure_velocity() const
+{
+  return matrix_free_data->get_quad_index(field + quad_index_p_u);
+}
+
+template<int dim, typename Number>
+unsigned int
+SpatialOperator<dim, Number>::get_quad_index_pressure() const
+{
+  return matrix_free_data->get_quad_index(field + quad_index_p);
+}
+
+template<int dim, typename Number>
+unsigned int
+SpatialOperator<dim, Number>::get_quad_index_velocity() const
+{
+  return matrix_free_data->get_quad_index(field + quad_index_u);
+}
+
+template<int dim, typename Number>
+std::shared_ptr<dealii::Mapping<dim> const>
+SpatialOperator<dim, Number>::get_mapping() const
+{
+  return mapping;
+}
+
+template<int dim, typename Number>
+dealii::FiniteElement<dim> const &
+SpatialOperator<dim, Number>::get_fe_p() const
+{
+  return *fe_p;
+}
+
+template<int dim, typename Number>
+dealii::FiniteElement<dim> const &
+SpatialOperator<dim, Number>::get_fe_u() const
+{
+  return *fe_u;
+}
+
+template<int dim, typename Number>
+dealii::DoFHandler<dim> const &
+SpatialOperator<dim, Number>::get_dof_handler_p() const
+{
+  return dof_handler_p;
+}
+
+template<int dim, typename Number>
+dealii::DoFHandler<dim> const &
+SpatialOperator<dim, Number>::get_dof_handler_u() const
+{
+  return dof_handler_u;
+}
+
+template<int dim, typename Number>
+dealii::AffineConstraints<Number> const &
+SpatialOperator<dim, Number>::get_constraint_p() const
+{
+  return constraint_p;
+}
+
+template<int dim, typename Number>
+dealii::AffineConstraints<Number> const &
+SpatialOperator<dim, Number>::get_constraint_u() const
+{
+  return constraint_u;
+}
+
+template<int dim, typename Number>
+dealii::types::global_dof_index
+SpatialOperator<dim, Number>::get_number_of_dofs() const
+{
+  return dof_handler_u.n_dofs() + dof_handler_p.n_dofs();
+}
+
+/*
+ * Initialization of vectors.
+ */
+template<int dim, typename Number>
+void
+SpatialOperator<dim, Number>::initialize_dof_vector(BlockVectorType & dst) const
+{
+  // pressure (1st block) + velocity (2nd block)
+  dst.reinit(2);
+
+  matrix_free->initialize_dof_vector(dst.block(0), get_dof_index_pressure());
+  matrix_free->initialize_dof_vector(dst.block(1), get_dof_index_velocity());
```

Comment:
I have seen these block-indices {0, 1} quite often in the code, also in the postprocessor. This could be somewhat error prone. We might have done it like this also in the past in other operators. How about introducing public static member variables `SpatialOperator::block_index_pressure` and `SpatialOperator::block_index_velocity`?

Reply:
Should this be defined in `SpatialOperator`? Dependent, where it is used, this requires "unnecessary" includes of the operator. 
What about introducing something similar to `numbers::invalid_timestep`, i.e. `Acoustics::numbers::block_index_p`?

Reply:
I don't think so, see also comment https://github.com/exadg/exadg/pull/592#discussion_r1390835643.

Reply:
`Operators::Operator` also needs this information. But from there, one could pass this information to `SpatialOperator`, where `Operators::Operator` is a member variable.

Reply:
It will also be needed in future convective operators, so I think it should be defined in the `SpatialOperator`. I restructured accordingly and used a forward declaration to avoid cyclic dependencies.

Reply:
Ok. Then it would be more natural in my opinion to inject these parameters into the operator via `OperatorData`. We follow a similar procedure in other modules where one `SpatialOperator` has different `Operators`, which have partial overlap in parameters. Note that all the `Operators` in ExaDG receive dof-index / quad-index in a very similar fashion from the `SpatialOperator` class. The current design with forward declaration is sub-optimal in  my opinion. It is also not really modular: Prinicipally, we want to be able write modified / extended / new `SpatialOperators` using existing `Operators`. With the present solution, `Operators` depend explicitly on one particular `SpatialOperator`, which is against the modular design of arranging existing `Operators` in another `SpatialOperator`.